extended-utxo-spec/extended-utxo-specification.tex

```diff
@@ -0,0 +1,1212 @@
+This section defines some basic notation.  We generally follow the
+notation established by \cite{Zahnentferner18-UTxO}.
+
+\begin{ruledfigure}{H}
+  \begin{itemize}
+\item Types are typeset in $\mathsf{sans~serif}$.
+
+\item All scripts are of type $\mathsf{Script}$.
+
+\item The only operation on $\mathsf{Script}$ from the ledger's
+  perspective is applying a validator script to three arguments,
+  which is denoted by $\llbracket \cdots \rrbracket :
+  \mathsf{Script} \rightarrow \mathsf{Script} \times \mathsf{Script}
+    \times \mathsf{Script} \rightarrow \mathbb{V}$ where $\mathbb{V}$
+    is some type containing a distinguished value
+    \textsf{error}\footnotemark.
+
+\item A record type with fields $\phi_1, \ldots, \phi_n$ of types $T_1,
+  \ldots, T_n$ is denoted by $(\phi_1 : T_1, \ldots, \phi_n : T_n)$.
+
+\item If $t$ is a value of a record type $T$ and $\phi$ is the name
+  of a field of $T$ then $t.\phi$ denotes the value of $\phi$ for
+  $t$.
+  
+\item A list $\lambda$ of type $\List{T}$ is
+    either the empty list $[]$ or a list $e :: \lambda'$ with $head$
+    $e$ of type $T$ and $tail$ $\lambda'$ of type
+    $\List{T}$. We denote the $i$th element of a list
+    $L$ by $L[i]$.
+    The concatenation of two lists $\lambda_1$ and
+    $\lambda_2$ is denoted $\lambda_1 ::: \lambda_2$.
+
+  \item For two types $K$ and $V$, $\Map{K}{V}$ denotes the type of
+    partial functions from \textit{keys} in $K$ to \textit{values} in
+    $V$.  If $m$ is such a map then $\dom m$ (the \textit{domain} of
+    $m$) is the set of keys of $m$ (ie, $K$).  Given a map $m$ and a
+    key $k \in \dom m$, we denote the result of looking up $k$ in $m$
+    by $m[k]$.  Equality for maps is defined in the obvious way: $m_1
+    = m_2$ if and only if $\dom m_1 = \dom m_2$ and $m_1[k] = m_2[k]$
+    for all $k \in \dom m_1$ (assuming we have decidable equality in $V$).
+    
+  \item If the type $V$ is a monoid with binary operation $+$ (for
+    example, a numeric type) and $f, g \in \Map{K}{V}$ then we define
+    their sum $f+g \in \Map{K}{V}$ with domain $\dom (f+g) = \dom f
+    \cup \dom g$ by
+    \[
+    (f+g)[k] =
+    \left\{ \begin{array}{ll}
+        f[k] + g[k] & \mbox{if $k \in \dom f \cap \dom g$}\\
+        f[k] & \mbox{if $k \in \dom f \setminus \dom g$}\\
+        g[k] & \mbox{if $k \in \dom g \setminus \dom f$}
+      \end{array}
+      \right.
+      \]
+      We generalise this to the sum of a finite number of maps using
+      the $\sum$ notation in the usual way.  Note that the type
+      $\Map{K}{V}$ itself becomes a monoid under the operation $+$
+      which we have just defined (with the empty map as identity element).
+
+    \item $x \mapsto f(x)$ denotes an anonymous function.
+
+  \item A cryptographic
+    collision-resistant hash of a value $c$ is denoted $c^{\#}$.
+
+  \item If any operation fails (for example, indexing a list out of
+    range or trying to spend an output which has already been spent)
```

Comment:
I think I put that in because it's kind of a mathematical specification, but there's a practical question about what happens if something's out of range. Usually in mathematics you can arrange that everything's well-defined in advance, so "run-time" errors like this don't arise.  I was trying to weasel my way out of threading errors though everything.

Reply:
The errors in scripts are conceptually very different from any errors that might arise during validation outside of scripts, and I don't think we should conflate them. I continue to think that we shouldn't need to talk about script errors at all.